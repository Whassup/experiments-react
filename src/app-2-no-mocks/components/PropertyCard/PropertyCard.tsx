// Exhibit A: Fat dependencies

import { FunctionComponent } from "react";

export interface PropertyCardProps {
  title: string;
  href: string;
  attributes?: React.ReactNode;
  onSaveToggle: () => void;
  saved?: boolean;
}

export const PropertyCard: FunctionComponent<PropertyCardProps> = ({
  title,
  attributes,
  href,
  saved = false,
  onSaveToggle,
}) => {
  return (
    <article>
      <h1>
        <a href={href}>{title}</a>
      </h1>
      {attributes}
      <a href={href}>View details page</a>
      <button type="button" onClick={onSaveToggle}>
        {saved ? "Remove" : " Save"}
      </button>
    </article>
  );
};
