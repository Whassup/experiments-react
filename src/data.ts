// TODO:  Add a property image to this example
// Add a results list with pageinfo

export interface PropertyData {
    property : {
        id: string,
        address: {
            shortAddressDisplay: string,
        },
        attributes: {
            beds: number,
            baths: number,
            carSpaces: number
        }
    },
    propertyDetailsPage: {
        slug: string
    }
}

export const createPropertyData = (): PropertyData =>({
    property : {
        id: String(Math.floor(Math.random() * 1000)),
        address: {
            shortAddressDisplay: "7 Westview Court, Springvale South",
        },
        attributes: {
            beds: 1,
            baths: 2,
            carSpaces: 3
        }
    },
    propertyDetailsPage: {
        slug: "details/7-westview-court-springvale-south-3172-vic"
    }
});


export interface PropertyResultsData {
    results: PropertyData[];
    pageInfo: {
        totalResultsCount: number;
        hasMore: boolean;
    }
}

export const createPropertyResultsData = (): PropertyResultsData =>({
    results: [createPropertyData(), createPropertyData(), createPropertyData()],
    pageInfo: {
        totalResultsCount: 100,
        hasMore: true
    }
});