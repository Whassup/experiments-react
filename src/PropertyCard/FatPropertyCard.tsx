// Exhibit A: Fat dependencies

import { FunctionComponent } from "react";
import { data } from "./data";


interface PropertyCardProps {
    data: typeof data;
}

export const FatPropertyCard: FunctionComponent<PropertyCardProps> = ({ data }) => {
    const { property, propertyDetailsPage } = data;
    const { address: { shortAddressDisplay }, attributes: { beds, baths, carSpaces } } = property;
    const { slug } = propertyDetailsPage;

    return <article>
        <h1><a href={slug}>{shortAddressDisplay}</a></h1>
        <div>
            <h2>Property Attributes</h2>
            <dl>
                <dt>Beds</dt>
                <dd aria-label="beds">{beds}</dd>

                <dt>Baths</dt>
                <dd aria-label="baths">{baths}</dd>

                <dt>Carspaces</dt>
                <dd aria-label="carSpaces">{carSpaces}</dd>
            </dl>
        </div>
        <a href={slug}>View details page</a>
    </article>
}


export default function FatPropertyCardWithData () {
    return <FatPropertyCard data={data} />
}