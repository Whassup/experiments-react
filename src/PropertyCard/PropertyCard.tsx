// Exhibit A: Fat dependencies

import { FunctionComponent } from "react";
import { data } from "./data";

interface PropertyAttributesProps {
    beds: number;
    baths: number;
    carSpaces: number;
}

export const PropertyAttributes: FunctionComponent<PropertyAttributesProps> = ({ beds, baths, carSpaces}) => (<div>
    <h2>Property Attributes</h2>
    <dl>
        <dt>Beds</dt>
        <dd aria-label="beds">{beds}</dd>

        <dt>Baths</dt>
        <dd aria-label="baths">{baths}</dd>

        <dt>Carspaces</dt>
        <dd aria-label="carSpaces">{carSpaces}</dd>
    </dl>
</div>);

interface PropertyCardProps {
    title: string;
    href: string;
    attributes: React.ReactNode;
}

export const PropertyCard: FunctionComponent<PropertyCardProps> = ({ title, attributes, href }) => {
    return <article>
        <h1><a href={href}>{title}</a></h1>
        { attributes }
        <a href={href}>View details page</a>
    </article>
}


interface PropertyCardWithDataProps {
    data: typeof data;
}
export const PropertyCardWithData: FunctionComponent<PropertyCardWithDataProps> = ({ data }) => {
    const { property, propertyDetailsPage } = data;
    const { address: { shortAddressDisplay }, attributes: { beds, baths, carSpaces } } = property;
    const { slug } = propertyDetailsPage;

    return <PropertyCard title={shortAddressDisplay} href={slug} attributes={<PropertyAttributes beds={beds} baths={baths} carSpaces={carSpaces} />} />
}