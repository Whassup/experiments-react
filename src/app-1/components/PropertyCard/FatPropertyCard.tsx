// Exhibit A: Fat dependencies

import { FunctionComponent } from "react";
import { PropertyData } from "../../../data";


interface FatPropertyCardProps {
    data: PropertyData;
    onSaveToggle: (propertyData: PropertyData) => void;
    saved?: boolean;
}

export const FatPropertyCard: FunctionComponent<FatPropertyCardProps> = ({ data, saved, onSaveToggle}) => {
    const { property, propertyDetailsPage } = data;
    const { address: { shortAddressDisplay }, attributes: { beds, baths, carSpaces } } = property;
    const { slug } = propertyDetailsPage;

    return <article data-testid={`card-${property.id}${saved ? "--saved" : ""}`} >
        <h1><a href={slug}>{shortAddressDisplay}</a></h1>
        <div>
            <h2>Property Attributes</h2>
            <dl>
                <dt>Beds</dt>
                <dd aria-label="beds">{beds}</dd>

                <dt>Baths</dt>
                <dd aria-label="baths">{baths}</dd>

                <dt>Carspaces</dt>
                <dd aria-label="carSpaces">{carSpaces}</dd>
            </dl>
        </div>
        <a href={slug}>View details page</a>
        <button data-testid={`save-${property.id}`} type="button" onClick={() => onSaveToggle(data)}>{  saved ? "Remove" : "Save" }</button>
    </article>
}

FatPropertyCard.defaultProps = {
    saved : false
}