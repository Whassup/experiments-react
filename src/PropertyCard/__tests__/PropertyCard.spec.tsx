import { render, screen } from "@testing-library/react";
import { createPropertyData } from "../data";
import { PropertyAttributes, PropertyCard, PropertyCardWithData } from "../PropertyCard";


describe("<PropertyCardWithData />", () => {
    it("SHOULD render with mapped data", () => {
        const { asFragment } = render(<PropertyCardWithData data={createPropertyData()} />);
        expect(asFragment()).toMatchSnapshot();
    });
});

describe("<PropertyCard />", () => {
    describe("render", () => {
        const title = "7 Westview Court Springvale South";
        const href = "http://localhost/7-westview-court-springvale-south";
        const attributesContent = "Beds 4, Baths 3, Carspaces 4";
        const Attributes = () => <>{attributesContent}</>;

        beforeEach(() => {
            render(<PropertyCard title={title} href={href} attributes={<Attributes />} />);
        });

        it("SHOULD display property card title", () => {
            const titleElement = screen.getByText(title);
            expect(titleElement).toBeInTheDocument();
        });

        it("title SHOULD link to href", () => {
            const link = screen.queryByRole("link", { name: title });
            const href = (link instanceof HTMLAnchorElement && link.href) ?? "";
            expect(href).toBe(href);
        });

        it("view details SHOULD link to href", () => {
            const link = screen.queryByRole("link", { name: "View details page" });
            const href = (link instanceof HTMLAnchorElement && link.href) ?? "";
            expect(href).toBe(href);
        });

        it("SHOULD display attributes", () => {
            const attributesElement = screen.getByText(attributesContent);
            expect(attributesElement).toBeInTheDocument();
        });
    });
});

describe("<PropertyAttributes />", () => {
    describe("render", () => {
        const beds = 0;
        const baths = 1;
        const carspaces = 2;
        beforeEach(() => {
            render(<PropertyAttributes beds={beds} baths={baths} carSpaces={carspaces} />);
        });

        it("SHOULD display beds count", () => {
            const bedsCount = screen.getByRole("definition", { name: "beds" });
            expect(bedsCount).toHaveTextContent(String(beds));
        });

        it("SHOULD display baths count", () => {
            const bathsCount = screen.getByRole("definition", { name: "baths" });
            expect(bathsCount).toHaveTextContent(String(baths));
        });

        it("SHOULD display carSpaces count", () => {
            const carSpacesCount = screen.getByRole("definition", { name: "carSpaces" });
            expect(carSpacesCount).toHaveTextContent(String(carspaces));
        });
    });
});