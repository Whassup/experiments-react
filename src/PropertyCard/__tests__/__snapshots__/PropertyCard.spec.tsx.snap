// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PropertyCardWithData /> SHOULD render with mapped data 1`] = `
<DocumentFragment>
  <article>
    <h1>
      <a
        href="details/7-westview-court-springvale-south-3172-vic"
      >
        7 Westview Court, Springvale South
      </a>
    </h1>
    <div>
      <h2>
        Property Attributes
      </h2>
      <dl>
        <dt>
          Beds
        </dt>
        <dd
          aria-label="beds"
        >
          1
        </dd>
        <dt>
          Baths
        </dt>
        <dd
          aria-label="baths"
        >
          2
        </dd>
        <dt>
          Carspaces
        </dt>
        <dd
          aria-label="carSpaces"
        >
          3
        </dd>
      </dl>
    </div>
    <a
      href="details/7-westview-court-springvale-south-3172-vic"
    >
      View details page
    </a>
  </article>
</DocumentFragment>
`;
